Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift

```diff
@@ -320,6 +335,33 @@ open class MobileSharedApplication: SharedApplication {
             }
         }
     }
+
+    public func presentLoading() {
+        if let safariViewController = controller?.presentedViewController as? MobileSafariViewController {
+            // Web OAuth flow, present the spinner over the MobileSafariViewController.
+            safariViewController.present(LoadingViewController(nibName: nil, bundle: nil), animated: false)
+        } else {
+            if let loadingStatusDelegate = loadingStatusDelegate {
+                loadingStatusDelegate.showLoading()
+            } else {
+                controller?.present(LoadingViewController(nibName: nil, bundle: nil), animated: false)
+            }
+        }
+    }
+
+    public func dismissLoading() {
+        if let safariViewController = controller?.presentedViewController as? MobileSafariViewController,
+            let loadingView = safariViewController.presentedViewController as? LoadingViewController {
+            // Web OAuth flow, dismiss loading view on the MobileSafariViewController.
+            loadingView.dismiss(animated: false)
+        } else {
+            if let loadingStatusDelegate = loadingStatusDelegate {
+                loadingStatusDelegate.dismissLoading()
+            } else if let loadingView = controller?.presentedViewController as? LoadingViewController {
+                loadingView.dismiss(animated: false)
+            }
+        }
```

Comment:
This logic is pretty hard to follow. Could we have a helper computed property that returns an optional loading controller ? That would would simplify logic like:

```
var loadingController: LoadingViewController? {
    let safariController = controller?.presentedViewController as? MobileSafariViewController
    return safariViewController.presentedViewController as? LoadingViewController ?? controller?.presentedViewController as? LoadingViewController
}
```

This method would then simplify to:


```
    if let loadingController = self.loadingController {
        loadingController.dismiss(animated: false)
    } else {
        loadingStatusDelegate?.dismissLoading()
    }
```

This is nicer as it's exhaustive and you dont have all these weird casting going on.

Reply:
yes, will make a helper to simplify the logics. thanks!